### DISCUSSION TITLE
Collecting feedback for custom layout guide

### DISCUSSION THREAD

**Discussion by redphx** [redphx/better-xcloud#285] [Custom touch controls]
If you have feedback or questions about [the guide](https://github.com/redphx/better-xcloud/wiki/How-to-create-custom-touch-control-layout) please post it here.  
You can also use this post to ask for help when designing a layout.

Thanks.

**Comment by Tak-attack** [comment]
I don't know if this will help, but I will share how I did it.

I had never worked with JSON files before using Better xcloud, but within a week or so I had a rough understanding and was able to create custom layouts for Halo Infinite and Shredders. (I knew some of the basics of programming).

- First, we tried various JSON editors to understand the structure of JSON files. As a result, I decided to use [JSON Editor Online](https://jsoneditoronline.org/).
It allows you to edit in a tree format and to beautify (indent with 2 spaces) and validate JSON content.
I also like the color picker available.

- I also think it would be very useful to create a custom layout blueprint in advance.
So I created a simple blueprint in a spreadsheet.
<img width="881" alt="スクリーンショット 2024-03-11 060639" src="https://github.com/redphx/better-xcloud/assets/149154024/e9dae8f7-e513-4205-bcb2-8c7d40313787">

This way, it is easy to see at a glance in which order to describe the objects.
![309552295-d5bcddaf-0423-49c9-87e8-67745da7eb38](https://github.com/redphx/better-xcloud/assets/149154024/c3b28dd3-a703-4b24-94c8-98a0a4a8e83d)

In fact, it would be very nice to have a tool that would allow me to just visually place a button and export it to a JSON file, but alas, I do not have such skills...

**Comment by Tak-attack** [comment]
> **Limitations**
> - Only supports one layout for the entire game (cannot dynamically switch layout based on context)

I think this part is inappropriate because one game can switch between multiple layouts.

- It is not possible to create a layout that automatically switches between native touch controls and situations.

Would the above wording be appropriate?

**Reply by redphx** [reply]
Thanks. I'll update it.

**Comment by Kode-Z** [comment]
After I’ve pasted the BX_EXPOSED.test_touch_control, I can easily edit the default controls fine. If I use anything other than that (for example, the layout from another game) the touch controls disappear? Am I missing something?

**Reply by redphx** [reply]
Which value did you pass to the `test_touch_control` function?

**Reply by Kode-Z** [reply]
I tested this layout so I can attempt to modify it.

https://github.com/redphx/better-xcloud/blob/gh-pages/touch-layouts/layouts/halo-infinite.json

**Reply by redphx** [reply]
Don't pass the entire layout in. Extract the value of `content` property only.

**Reply by Kode-Z** [reply]
Can you please explain it with an example?

**Reply by redphx** [reply]
Use this part only

<img width="391" alt="image" src="https://github.com/redphx/better-xcloud/assets/96280/e3c58062-35ff-45c2-a9c4-b000a3f4b73b">

If your text editor supports JSON you can easily collapse the node and select its content:

<img width="454" alt="image" src="https://github.com/redphx/better-xcloud/assets/96280/08faf8f5-74ca-419f-b382-1d40e62b5184">

I'm using Sublime Text -> Cmd + Shift + P -> `Set syntax: JSON`

**Reply by Kode-Z** [reply]
![IMG_9689](https://github.com/redphx/better-xcloud/assets/162533196/802b11f2-6dce-4a1d-991a-d2dc19cc29fb)

Made some progress! Just missing the D pad!

**Reply by Kode-Z** [reply]
I’ve tried adding the D pad, but had no luck with it.

**Reply by redphx** [reply]
Did you check the docs? https://learn.microsoft.com/en-us/gaming/gdk/_content/gc/reference/system/touchadaptationkit/controls/game-streaming-touch-directionalpad

**Reply by Kode-Z** [reply]
I did, but I don’t understand where to implement the code.

**Reply by Kode-Z** [reply]
![IMG_9690](https://github.com/redphx/better-xcloud/assets/162533196/8c8cc95d-1964-4597-a96c-b33c20a56342)

Would it be around here somewhere?

**Reply by redphx** [reply]
You can use the default layout in the guide for reference

**Reply by Kode-Z** [reply]
![IMG_9692](https://github.com/redphx/better-xcloud/assets/162533196/2cfee06b-8c95-4edb-9659-31326cdbe90d)

I added this line (to the left stick), but nothing is showing. (I tried to follow the default layout).

**Reply by redphx** [reply]
It's not showing because it's out of view. Try to add a few more `null` (or button) before it.

**Reply by Kode-Z** [reply]
I’ve been trying for a while, but I can’t get the dpad working unfortunately.

Here is the script:
```javascript
BX_EXPOSED.test_touch_control(
    {
    "left": {
      "inner": [
        {
          "type": "joystick",
          "axis": {
            "input": "axisXY",
            "output": "leftJoystick"
          },
          "styles": {
            "default": {
              "knob": {
                "faceImage": {
                  "type": "icon",
                  "value": "walk"
                }
              }
            }
          }
        }
      ]
    },
    "right": {
      "inner": [
        {
          "type": "joystick",
          "axis": {
            "input": "axisXY",
            "output": "rightJoystick"
          },
          "styles": {
            "default": {
              "knob": {
                "faceImage": {
                  "type": "icon",
                  "value": "look"
              }
            }
          }
        }
       }
      ],
      "outer": [
        [
          {
            "type": "button",
            "action": "leftTrigger",
            "toggle": true,
            "styles": {
              "default": {
                "faceImage": {
                  "type": "icon",
                  "value": "aim"
                }
              }
            }
          },
          {
          "type": "button",
          "action": "leftThumb",
          "styles": {
            "default": {
              "faceImage": {
                "type": "icon",
                "value": "crouch"
              }
            }
          }
        },
        ],
        {
          "type": "button",
          "action": "rightBumper",
          "styles": {
            "default": {
              "faceImage": {
                "type": "icon",
                "value": "fire"
              }
            }
          }
        },
        {
          "type": "button",
          "action": "gamepadA",
          "styles": {
            "default": {
              "faceImage": {
                "type": "icon",
                "value": "interact"
              }
            }
          }
        },
        [
          {
            "type": "touchpad",
            "axis": {
              "input": "axisXY",
              "output": "rightJoystick",
              "deadzone": {
                "threshold": 0.5,
                "radial": true
              }
            },
            "renderAsButton": true,
            "action": [
              "rightTrigger"
            ],
            "styles": {
              "default": {
                "faceImage": {
                  "type": "icon",
                  "value": "fire"
                }
              }
            }
          },
          {
            "type": "button",
            "action": "gamepadB",
            "styles": {
              "default": {
                "faceImage": {
                  "type": "icon",
                  "value": "punch"
                }
              }
            }
          },
          {
            "type": "button",
            "action": "gamepadX",
            "styles": {
              "default": {
                "faceImage": {
                  "type": "icon",
                  "value": "reload"
                }
              }
            }
          }
        ],
        {
          "type": "button",
          "action": "leftBumper",
          "styles": {
            "default": {
              "faceImage": {
                "type": "icon",
                "value": "roll"
              }
            }
          }
        },
        {
          "type": "button",
          "action": "gamepadY",
          "styles": {
            "default": {
              "faceImage": {
                "type": "icon",
                "value": "weaponSelect"
              }
            }
          }
        }
      ]
    },
    
    "upper": {
      "right": [
        {
          "type": "button",
          "action": "menu"
        },
        {
          "type": "button",
          "action": "view"
        }
      ]
    },
  }
)
```

**Reply by redphx** [reply]
Put this in the `left` object:  

```json
"outer": [
  null,
  null,
  null,
  {
      "type": "directionalPad",
      "scale": 1.0
  }
]
```

<img width="356" alt="image" src="https://github.com/redphx/better-xcloud/assets/96280/9d766a5f-2c18-4b87-bcbc-4a8aafbe5c92">

**Reply by Kode-Z** [reply]
Thanks. I tried it, but it’s still not showing the D Pad for some reason. Thanks for all the assistance, by the way. Really appreciate it!

![IMG_9693](https://github.com/redphx/better-xcloud/assets/162533196/2f2c885c-9bb9-45ad-af97-cebed930db5d)

**Reply by Kode-Z** [reply]
Everything else is as I want it, just need the DPad for completion.

**Reply by Kode-Z** [reply]
![IMG_9694](https://github.com/redphx/better-xcloud/assets/162533196/f528f3dc-3e4f-423f-baec-f343f632db3c)

**Reply by redphx** [reply]
I said put it "in the `left` object", not "next to"

**Comment by redphx** [comment]
@Tak-attack @Kode-Z @evanpaul14 the official tool to create touch control has been released by xCloud team: https://github.com/microsoft/xbox-game-streaming-tools/releases/tag/tak-cli-v11.0.0  

My laptop is broken so I can't check this right now.

**Reply by evanpaul14** [reply]
I used it for #324. The ability to quickly test new layouts and the premade templates make the process a lot faster overall.
![Screenshot 2024-03-26 at 8 06 13 PM](https://github.com/redphx/better-xcloud/assets/120766197/5e4b83be-90df-4971-8c3d-550a279906dc)

**Comment by redphx** [comment]
I updated the guide with info about the TAK extension for VS Code.  

I also changed param of `BX_EXPOSED.test_touch_control()`.

- **Before:** you pass only the "content" value of the layout.  
- **Now:** you pass the entire JSON layout created by TAK extension.

It now supports ["styles" property](https://learn.microsoft.com/en-us/gaming/gdk/_content/gc/reference/system/touchadaptationkit/layout/game-streaming-touch-layout-styles). Use this to reuse colors in the layout.
